src/autorest-core/language-service/language-service.ts

```diff
@@ -0,0 +1,653 @@
+    const refValueJsonPath = docAnalysis.GetJsonPathFromJsonReferenceAt(position);
+    if (refValueJsonPath) {
+
+      for (const location of docAnalysis.GetDefinitionLocations(refValueJsonPath)) {
+        yield {
+          language: "yaml",
+          value: safeDump(location.value)
+        };
+      }
+    } // else {console.log("found nothing that looks like a JSON reference"); return null; }
+  }
+
+  private * onHoverJsonPath(docAnalysis: DocumentAnalysis, position: Position): Iterable<MarkedString> {
+    const potentialQuery: string = <string>docAnalysis.GetJsonQueryAt(position);
+    if (potentialQuery) {
+
+      const queryNodes = [...docAnalysis.GetDefinitionLocations(potentialQuery)];
+      yield {
+        language: "plaintext",
+        value: `${queryNodes.length} matches\n${queryNodes.map(node => node.jsonPath).join("\n")}`
+      };
+    } // else { console.log("found nothing that looks like a JSON path"); return null; }
+  }
+
+
+  private async onHover(position: TextDocumentPositionParams): Promise<Hover> {
+    const docAnalysis = await this.getDocumentAnalysis(position.textDocument.uri);
+    return docAnalysis ? <Hover>{
+      contents: [
+        ...this.onHoverRef(docAnalysis, position.position),
+        ...this.onHoverJsonPath(docAnalysis, position.position)
+      ]
+    } : <Hover><any>null
+  }
+
+  private onDefinitionRef(docAnalysis: DocumentAnalysis, position: Position): Iterable<Location> {
+    const refValueJsonPath = docAnalysis.GetJsonPathFromJsonReferenceAt(position);
+    if (refValueJsonPath) {
+      return docAnalysis.GetDocumentLocations(refValueJsonPath);
+    } // else  { console.log("found nothing that looks like a JSON reference"); }
+    return [];
+  }
+
+  private onDefinitionJsonPath(docAnalysis: DocumentAnalysis, position: Position): Iterable<Location> {
+    const potentialQuery: string = <string>docAnalysis.GetJsonQueryAt(position);
+    if (potentialQuery) {
+
+      return docAnalysis.GetDocumentLocations(potentialQuery);
+    } // else  { console.log("found nothing that looks like a JSON path");}
+    return [];
+  }
+
+  private async onDefinition(position: TextDocumentPositionParams): Promise<Location[]> {
+    const docAnalysis = await this.getDocumentAnalysis(position.textDocument.uri);
+    return docAnalysis ? [
+      ...this.onDefinitionRef(docAnalysis, position.position),
+      ...this.onDefinitionJsonPath(docAnalysis, position.position)
+    ] : [];
+  }
+
+  private async onFileEvents(changes: FileEvent[]) {
+    this.debug(`onFileEvents: ${changes}`);
+    for (const each of changes) {
+
+      const doc = this.get(each.uri);
+      if (doc) {
+        this.onDocumentChanged(doc);
+        return;
+      }
+
+      let documentUri = each.uri;
+      const txt = await this.ReadFile(each.uri);
+      if (documentUri.startsWith("file://")) {
+        // fake out a document for us to play with
```

Comment:
what case does this correspond to? just opening a file and start typing something?